File: lib/web_ui/lib/src/engine/compositor/vertices.dart
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.


part of engine;

js.JsArray<Float32List> _encodeRawColorList(Int32List rawColors) {
  final int colorCount = rawColors.length;
  final List<ui.Color> colors = <ui.Color>[];
  for (int i = 0; i < colorCount; ++i) {
    colors.add(ui.Color(rawColors[i]));
  }
  return makeColorList(colors);
}

class CkVertices implements ui.Vertices {
  js.JsObject? skVertices;

  CkVertices(
    ui.VertexMode mode,
    List<ui.Offset> positions, {
    List<ui.Offset>? textureCoordinates,
    List<ui.Color>? colors,
    List<int>? indices,
  })  : assert(mode != null), // ignore: unnecessary_null_comparison
        assert(positions != null) { // ignore: unnecessary_null_comparison
    if (textureCoordinates != null &&
        textureCoordinates.length != positions.length)
      throw ArgumentError(
          '"positions" and "textureCoordinates" lengths must match.');
    if (colors != null && colors.length != positions.length)
      throw ArgumentError('"positions" and "colors" lengths must match.');
    if (indices != null &&
        indices.any((int i) => i < 0 || i >= positions.length))
      throw ArgumentError(
          '"indices" values must be valid indices in the positions list.');

    final js.JsArray<js.JsArray<double>>? encodedPositions = encodePointList(positions);
    final js.JsArray<js.JsArray<double>>? encodedTextures =
        encodePointList(textureCoordinates);
    final js.JsArray<Float32List>? encodedColors =
        colors != null ? makeColorList(colors) : null;
    if (!_init(mode, encodedPositions, encodedTextures, encodedColors, indices))
      throw ArgumentError('Invalid configuration for vertices.');
  }

  CkVertices.raw(
    ui.VertexMode mode,
    Float32List positions, {
    Float32List? textureCoordinates,
    Int32List? colors,
    Uint16List? indices,
  })  : assert(mode != null), // ignore: unnecessary_null_comparison
        assert(positions != null) { // ignore: unnecessary_null_comparison
    if (textureCoordinates != null &&
        textureCoordinates.length != positions.length)
      throw ArgumentError(
          '"positions" and "textureCoordinates" lengths must match.');
    if (colors != null && colors.length * 2 != positions.length)
      throw ArgumentError('"positions" and "colors" lengths must match.');
    if (indices != null &&
        indices.any((int i) => i < 0 || i >= positions.length))
      throw ArgumentError(
          '"indices" values must be valid indices in the positions list.');

    if (!_init(
      mode,
      encodeRawPointList(positions) as js.JsArray<js.JsArray<double>>?,
      encodeRawPointList(textureCoordinates) as js.JsArray<js.JsArray<double>>?,
      colors != null ? _encodeRawColorList(colors) : null,
      indices,
    )) {
      throw ArgumentError('Invalid configuration for vertices.');
    }
  }

  bool _init(
      ui.VertexMode mode,
      js.JsArray<js.JsArray<double>>? positions,
      js.JsArray<js.JsArray<double>>? textureCoordinates,
      js.JsArray<Float32List>? colors,
      List<int>? indices) {
    js.JsObject? skVertexMode;
    switch (mode) {
      case ui.VertexMode.triangles:
        skVertexMode = canvasKit['VertexMode']['Triangles'];
        break;
      case ui.VertexMode.triangleStrip:
        skVertexMode = canvasKit['VertexMode']['TrianglesStrip'];
        break;
      case ui.VertexMode.triangleFan:
        skVertexMode = canvasKit['VertexMode']['TriangleFan'];
        break;
    }

    final js.JsObject? vertices =
        canvasKit.callMethod('MakeSkVertices', <dynamic>[
      skVertexMode,
      positions,
      textureCoordinates,
      colors,
      indices,
    ]);

    if (vertices != null) {
      skVertices = vertices;
      return true;
    } else {
      return false;
    }
  }
}

File: engine/src/flutter/impeller/geometry/vertices.cc
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "vertices.h"

namespace impeller {

Vertices::Vertices(std::vector<Point> points,
                   std::vector<uint16_t> indices,
                   std::vector<Color> colors,
                   VertexMode vertex_mode,
                   Rect bounds)
    : positions_(std::move(points)),
      indices_(std::move(indices)),
      colors_(std::move(colors)),
      vertex_mode_(vertex_mode),
      bounds_(bounds) {
  NormalizeIndices();
}

Vertices::~Vertices() = default;

bool Vertices::IsValid() const {
  size_t points_size = positions_.size();
  size_t colors_size = colors_.size();

  if (colors_size > 0 && colors_size != points_size) {
    return false;
  }

  return true;
}

std::optional<Rect> Vertices::GetBoundingBox() const {
  return bounds_;
};

std::optional<Rect> Vertices::GetTransformedBoundingBox(
    const Matrix& transform) const {
  auto bounds = GetBoundingBox();
  if (!bounds.has_value()) {
    return std::nullopt;
  }
  return bounds->TransformBounds(transform);
};

const std::vector<Point>& Vertices::GetPositions() const {
  return positions_;
}

const std::vector<uint16_t>& Vertices::GetIndices() const {
  return indices_;
}

const std::vector<Color>& Vertices::GetColors() const {
  return colors_;
}

VertexMode Vertices::GetMode() const {
  return vertex_mode_;
}

void Vertices::NormalizeIndices() {
  if (indices_.size() != 0 || positions_.size() == 0) {
    return;
  }
  for (size_t i = 0; i < positions_.size(); i++) {
    indices_.push_back(i);
  }
}

}  // namespace impeller
